[3] Add dropdown question type

TaylorAbraham opened the pull request:
Fixes #3 
* Add the dropdown question type which supports multiple options selected from a dropdown
* Updated unit tests to include the dropdown question type
* Added frontend code for this question type

![image](https://user-images.githubusercontent.com/7512973/77133091-364dc280-6a38-11ea-8189-17ba99aa4b62.png)

![image](https://user-images.githubusercontent.com/7512973/77133102-3cdc3a00-6a38-11ea-8bd6-81e04f709ccc.png)

The options are comma separated, which is explained in the tooltip:
![image](https://user-images.githubusercontent.com/7512973/77133271-e4f20300-6a38-11ea-91da-65058bc5a68f.png)

TaylorAbraham committed:
[3] Add backend groundwork for option question

TaylorAbraham committed:
Merge branch 'master' into 3/add-dropdown-question

TaylorAbraham committed:
[3] Commit updated package and package-lock

TaylorAbraham committed:
[3] Convert Collection to ArrayList for proper tables

TaylorAbraham committed:
[3] Add unit tests for dropdown question

MichaelDysart commented on src/main/java/webApp/webController.java:
```diff
@@ -53,6 +54,8 @@ public Response createSurvey(@RequestBody SurveyMessage surveyMessage) {
                             "Min is greater than max for question \"" + questionMsg.getQuestion() + "\"");
                 }
+            } else if (questionMsg.getType().equals(dropdownQuestionType)) {
+                questionList.add(new DropdownQuestion(questionMsg.getQuestion(), (ArrayList<String>) questionMsg.getOptions()));
```
If you are always casting to an ArrayList<String> then shouldn't `getOptions` just return an ArrayList<String>

MichaelDysart replied:
Otherwise if someone gets the wrong type in there this will crash.

TaylorAbraham replied:
Good catch, on it.

MichaelDysart commented on src/main/js/App.js:
```diff
@@ -72,6 +73,9 @@ const App = () => {
                 ]);
                 break;
+            case qType.DROPDOWN:
+                //TODO:
```
When will this be done?

TaylorAbraham replied:
Outdated, it has now been implemented.

TaylorAbraham committed:
[3] Add frontend for dropdown question

TaylorAbraham committed:
Merge branch 'master' into 3/add-dropdown-question

TaylorAbraham committed:
[3] Add back label for questions

TaylorAbraham committed:
[3] Fix broken tests

JacksonSchoe commented on src/main/java/webApp/QuestionMessage.java:
```diff
@@ -20,14 +23,16 @@
      * @param numberAnswer   the answer if the type is numerical
      */
-    public QuestionMessage(String type, String question, int min, int max, String stringAnswer, int numberAnswer, Collection<String> stringAnswerList, Collection<Integer> numberAnswerList) {
+    public QuestionMessage(String type, String question, Collection<String> options, int min, int max, String stringAnswer, int numberAnswer, Collection<String> stringAnswerList, Collection<Integer> numberAnswerList) {
```
We really should figure out a better way of going about this at some point, oh and my unit test involving this class will need to be updated to accommodate this change. It's under QuestionMessageTest.java. It will break with this change.

TaylorAbraham replied:
For sure agree. And yeah as you mentioned, tests have been fixed.

JacksonSchoe commented on src/main/java/webApp/QuestionMessage.java:
```diff
@@ -52,6 +57,12 @@ public String getQuestion() {
+     * Retrieve the question options
+     * @returns {Collection<String>}
+     */
+    public Collection<String> getOptions() { return options; }
```
This will also need to be tested in the test class I mentioned above.

TaylorAbraham replied:
On it.

JacksonSchoe commented on src/test/java/QuestionMessageTest.java:
```diff
@@ -27,8 +27,8 @@ public void setup() throws Exception{
 
-        qm1 = new QuestionMessage("Number","First 3 numbers?",1,3,"1,2,3", 123,sal1,nal1 );
-        qm2 = new QuestionMessage("String", "How to say bye?", 0,0,"Bye, See ya, Ciao", 0,sal2,nal2);
+        qm1 = new QuestionMessage("Number","First 3 numbers?", null, 1,3,"1,2,3", 123,sal1,nal1 );
```
Alright disregard my previous comments on these tests, you were clearly one step ahead of me

JacksonSchoe reviewed (approved):
Disregard my previous comments on the tests needing to be fixed, as you fixed them while I was writing the comment. Overall, good work. No complaints from my end

MichaelDysart commented on src/main/js/App.js:
```diff
@@ -108,7 +122,13 @@ const App = () => {
-            questions,
+            questions: questions.map(q => {
+                if (q.type === 'dropdown') {
+                    // For dropdowns, split the input before sending it
```
Ideally, we wouldn't be separating by comma because it prevents someone from inputting an option that contains a comma.
For now it is fine though.

TaylorAbraham replied:
Yeah I do agree, a future optimization could be to change the way we do this but I think this is the cleanest UI mechanism for the time being.

MichaelDysart committed:
Updated schema image

MichaelDysart committed:
Updated schema image

MichaelDysart committed:
Merge branch 'master' of https://github.com/MichaelDysart/SYSC4806_Project into 3/add-dropdown-question

 Conflicts:
	src/main/js/App.js
	src/test/java/webApp/SurveyMessageTest.java
	src/test/java/webApp/webControllerTest.java

MichaelDysart committed:
Fixed merge conflicts

MichaelDysart commented on src/main/js/App.js:
```diff
@@ -330,6 +344,43 @@ const App = () => {
+                                            color="primary"
+                                            onClick={() => deleteQuestion(i)}
+                                        ><DeleteIcon /></Button>
+                                        <Tooltip title="Separate dropdown options using commas">
```
Tooltip is not defined in this file

MichaelDysart commented on src/main/js/App.js:
```diff
@@ -330,6 +344,43 @@ const App = () => {
                                         ><DeleteIcon /></Button>
                                     </div>
                                 );
+                            case qType.DROPDOWN:
```
qType.DROPDOWN is not defined.

MichaelDysart commented on src/main/js/App.js:
```diff
@@ -330,6 +344,43 @@ const App = () => {
+                                            onClick={() => deleteQuestion(i)}
+                                        ><DeleteIcon /></Button>
+                                        <Tooltip title="Separate dropdown options using commas">
+                                            <HelpIcon />
```
HelpIcon is not defined in this file

TaylorAbraham replied:
Yeah it seemed like GitHub didn't update the diff correctly. As I'm sure you can see now, all of these constants are now defined or imported.

MichaelDysart replied:
Yes. I added them this morning.

MichaelDysart commented on src/main/js/App.js:
```diff
@@ -117,7 +128,13 @@ const App = () => {
+            questions: questions.map(q => {
+                if (q.type === 'dropdown') {
+                    // For dropdowns, split the input before sending it
+                    q.options = q.options.split(',');
```
When you call split here you replace q.options with an array. Thus, if you try to create another survey with the same options it will crash because split can't be called on an array.

TaylorAbraham replied:
I made sure to be careful of this. I am modifying a copy of the options and sending that to the server, but the state of the app is still left as an un-split array.

MichaelDysart replied:
This actually caused an error. All of the things I am pointing out I discovered because errors occurred.

MichaelDysart replied:
It wasn't modifying a copy of the options when I tested. It modified the actual options.

MichaelDysart committed:
Fixed bugs.
Added documentation.
Added unit tests.
Ensured 100% code coverage.
Improved formatting.

Fahid-Mannan commented on README.md:
```diff
@@ -10,6 +10,43 @@ create and register surveys, and collect results from the surveys.
+  
+  6. Added integration tests for controller.
+  
+  7. Added client side tests.
```
8. Updated all documentation including schema and UML

Fahid-Mannan commented on src/main/java/survey/DropdownQuestion.java:
```diff
@@ -0,0 +1,58 @@
+        this.options = options;
+    }
+
+    public ArrayList<String> getAnswers() {
```
This method and the previous couple are missing the Java doc comments

TaylorAbraham replied:
Good catch, on it.

Fahid-Mannan commented on src/main/java/webApp/webController.java:
```diff
@@ -81,10 +84,13 @@ public SurveyMessage retrieveSurvey(@RequestParam (name="id") int id) {
             for (Question question : survey.get().getQuestions()) {
                 if(question instanceof OpenEndedQuestion) {
```
For later refactor, a switch-case statement is far for efficient for cases like theses

TaylorAbraham replied:
Good note.

MichaelDysart replied:
Are you sure that a switch statement can be used for instanceof checking?
My reading of these comments suggests it is not https://stackoverflow.com/questions/5579309/is-it-possible-to-use-the-instanceof-operator-in-a-switch-statement

MichaelDysart replied:
Apparently you can do a switch described in here https://stackoverflow.com/questions/36332665/how-to-use-instanceof-in-a-switch-statement.
```
switch (true) {
    case error instanceof NotAllowedError: 
        return res.send(400);

    case error instanceof DatabaseEntryNotFoundError: 
        return res.send(404);

    default:
        log('Failed to do something async with an unspecified error: ', error);
        return res.send(500);
}
```
But has this actually improved anything? The switch true would likely look convoluted to someone unfamiliar with the code.

TaylorAbraham replied:
This will be useful to keep in mind for the future.

Fahid-Mannan reviewed (approved):
Overall well done, just some very minor feedback. I really like how the PR description is done
